netplan/libnetplan.py

```diff
@@ -18,19 +18,165 @@
 from typing import List, Union, IO
+from enum import IntEnum
 from io import StringIO
 import os
+import re
 
 
-class LibNetplanException(Exception):
+SCHEMA_PARSER_ERROR_MSG_REGEX = (
+    r'(?P<file_path>.*):(?P<error_line>\d+):(?P<error_col>\d+): (?P<message>.*)'
+)
+
+SCHEMA_VALIDATION_ERROR_MSG_REGEX = (
+    r'(?P<file_path>.*\.yaml): (?P<message>.*)'
+)
+
+
+class NETPLAN_ERROR_DOMAINS(IntEnum):
+    NETPLAN_PARSER_ERROR = 1
+    NETPLAN_VALIDATION_ERROR = 2
+    NETPLAN_FILE_ERROR = 3
+    NETPLAN_BACKEND_ERROR = 4
+    NETPLAN_EMITTER_ERROR = 5
+    NETPLAN_FORMAT_ERROR = 6
+
+
+class NETPLAN_PARSER_ERRORS(IntEnum):
+    NETPLAN_ERROR_INVALID_YAML = 0
+    NETPLAN_ERROR_INVALID_CONFIG = 1
+
+
+class NETPLAN_VALIDATION_ERRORS(IntEnum):
+    NETPLAN_ERROR_CONFIG_GENERIC = 0
+    NETPLAN_ERROR_CONFIG_VALIDATION = 1
+
+
+class NETPLAN_BACKEND_ERRORS(IntEnum):
+    NETPLAN_ERROR_UNSUPPORTED = 0
+    NETPLAN_ERROR_VALIDATION = 1
+
+
+class NETPLAN_EMITTER_ERRORS(IntEnum):
+    NETPLAN_ERROR_YAML_EMITTER = 0
+
+
+class NETPLAN_FORMAT_ERRORS(IntEnum):
+    NETPLAN_ERROR_FORMAT_INVALID_YAML = 0
+
+
+class NetplanException(Exception):
+    def __init__(self, message=None, domain=None, error=None):
+        self.domain = domain
+        self.error = error
+        self.message = message
+
+    def __str__(self):
+        return self.message
+
+
+class NetplanFileException(NetplanException):
     pass
 
 
-class _GError(ctypes.Structure):
+class NetplanValidationException(NetplanException):
+    '''
+    Netplan Validation errors are expected to contain the YAML file name
+    from where the error was found.
+    '''
+
+    def __init__(self, message=None, domain=None, error=None):
+        schema_error = re.match(SCHEMA_VALIDATION_ERROR_MSG_REGEX, message)
+        self.domain = domain
+        self.error = error
+        self.filename = schema_error["file_path"]
+        self.message = schema_error["message"]
+
+
+class NetplanParserException(NetplanException):
+    '''
+    Netplan Parser errors are expected to contain the YAML file name
+    and line and column numbers from where the error was found.
+    '''
+
+    def __init__(self, message=None, domain=None, error=None):
+        # Parser errors from libnetplan have the form:
+        #
+        # filename.yaml:4:14: Error in network definition: invalid boolean value 'falsea'
+        #
+        schema_error = re.match(SCHEMA_PARSER_ERROR_MSG_REGEX, message)
+
+        file_path = schema_error["file_path"]
+        error_line = schema_error["error_line"]
+        error_col = schema_error["error_col"]
+        error_message = schema_error["message"]
+
+        # Some error messages might have line breaks, let's ignore them.
+        # After removing the line breaks the error message might be left with a leading ':'
+        # which we also remove
+        msg = error_message.split('\n')
+        msg = msg[0].strip().strip(':')
```

Comment:
well, as far as I can see, the information we return after line breaks are purely for visualization, like:

```
/tmp/fakeroot/etc/netplan/test.yaml:4:7: Error in network definition: unknown key 'dhcp'
      dhcp: true
      ^
```

As we already give the caller the line and column numbers, I though it could be discarded. Do you have any example where we shouldn't do that?

Reply:
The only case I can currently find (`error.*\\n[^"]` to the rescue) is in `dbus.c`, which is not handled by this error-path... But what if we introduce some multi-line errors in the future? We would loose information.

Do we necessarily need to drop the error context here, or is it just "nicer looking"?

Reply:
I changed the regex to match everything :)